src/Bicep.Core/Registry/OciModuleRegistry.cs

```diff
@@ -132,27 +132,26 @@ public override async Task<bool> CheckArtifactExists(OciModuleReference referenc
             return true;
         }
 
-        private readonly ImmutableArray<string> allowedArtifactMediaTypes = ImmutableArray.Create(
-            BicepMediaTypes.BicepModuleArtifactType,
-            BicepMediaTypes.BicepProviderArtifactType);
-
-        private readonly ImmutableArray<string> allowedConfigMediaTypes = ImmutableArray.Create(
-            BicepMediaTypes.BicepModuleConfigV1,
-            BicepMediaTypes.BicepProviderConfigV1);
-
-        private readonly ImmutableArray<string> allowedMainLayerMediaTypes = ImmutableArray.Create(
-            BicepMediaTypes.BicepModuleLayerV1Json,
-            BicepMediaTypes.BicepProviderArtifactLayerV1TarGzip);
-
-        private (string mediaType, BinaryData data) getMainLayer(OciArtifactResult result)
+        private OciArtifactLayer getMainLayer(OciArtifactResult result)
         {
             if (result.Layers.Count() == 0)
             {
                 throw new InvalidModuleException("Expected at least one layer in the OCI artifact.");
             }
 
+            var expectedMediaType = result.Manifest.ArtifactType switch
+            {
+                // (Bicep v0.20.0 and lower use null for this field, so assume valid in that case)
+                null or BicepMediaTypes.BicepModuleArtifactType => BicepMediaTypes.BicepModuleLayerV1Json,
+                BicepMediaTypes.BicepProviderArtifactType => BicepMediaTypes.BicepProviderArtifactLayerV1TarGzip,
+                _ =>
+                     throw new InvalidModuleException(
+                            $"Unknown ArtifactType: '{result.Manifest.ArtifactType}'. Supported OCI artifactType fields are: (1) null or '{BicepMediaTypes.BicepModuleArtifactType}' for modules, or (2) '{BicepMediaTypes.BicepProviderArtifactType} for resource type providers'",
+                            InvalidModuleExceptionKind.WrongArtifactType)
```

Comment:
[nit]: This same exception (with the same message) is thrown both here and on line 176; you could add a method to throw it so that you don't need to declare the desired message twice.